The translation from assign to always_comb resulted in an error

I am a beginner, and I am using Verilog to write a simple RAM module:
```
module RAM (
    i_ram_ena  : input logic   ,
    i_ram_read : input logic   ,
    i_ram_write: input logic   ,
    i_ram_addr : input logic<8>,
    io_ram_data: inout logic<8>,
) {
    var memory: logic<8> [256];
    assign io_ram_data = if i_ram_read && i_ram_ena {
        memory[i_ram_addr]
    } else {
        8'hzz
    };

    always_ff (i_ram_write) {
        if i_ram_ena {
            memory[i_ram_addr] = io_ram_data;
        }
    }
}
```
This code was translated into:
```verilog
module RAM (
    input logic         i_ram_ena  ,
    input logic         i_ram_read ,
    input logic         i_ram_write,
    input logic [8-1:0] i_ram_addr ,
    inout logic [8-1:0] io_ram_data
);
    logic [8-1:0] memory      [0:256-1];
    always_comb io_ram_data = ((i_ram_read && i_ram_ena) ? (
        memory[i_ram_addr]
    ) : (
        8'hzz
    ));

    always_ff @ (posedge i_ram_write) begin
        if (i_ram_ena) begin
            memory[i_ram_addr] <= io_ram_data;
        end
    end
endmodule
```
When I use iverilog, I get the following error:
src/ram.sv:9: error: io_ram_data is not a valid l-value in RAM.
src/ram.sv:6:      : io_ram_data is declared here as wire.
However, when I change always_comb to assign, I don't get an error.
So, are there certain situations where assign should not be translated to always_comb?

--- a/crates/analyzer/src/analyzer_error.rs
+++ b/crates/analyzer/src/analyzer_error.rs
@@ -310,6 +310,20 @@ pub enum AnalyzerError {
 
     #[diagnostic(
         severity(Error),
+        code(missing_tri),
+        help("add tri type modifier"),
+        url("https://doc.veryl-lang.org/book/06_appendix/02_semantic_error.html#missing_tri")
+    )]
+    #[error("tri type modifier is required at inout port")]
+    MissingTri {
+        #[source_code]
+        input: NamedSource,
+        #[label("Error location")]
+        error_location: SourceSpan,
+    },
+
+    #[diagnostic(
+        severity(Error),
         code(too_large_enum_variant),
         help(""),
         url("https://doc.veryl-lang.org/book/06_appendix/02_semantic_error.html#too_large_enum_variant")
@@ -771,6 +785,13 @@ impl AnalyzerError {
             input: AnalyzerError::named_source(source, token),
             error_location: token.into(),
             reset: reset.into(),
+        }
+    }
+
+    pub fn missing_tri(source: &str, token: &TokenRange) -> Self {
+        AnalyzerError::MissingTri {
+            input: AnalyzerError::named_source(source, token),
+            error_location: token.into(),
         }
     }
 

--- a/crates/analyzer/src/handlers/check_direction.rs
+++ b/crates/analyzer/src/handlers/check_direction.rs
@@ -28,6 +28,31 @@ impl<'a> Handler for CheckDirection<'a> {
 }
 
 impl<'a> VerylGrammarTrait for CheckDirection<'a> {
+    fn port_declaration_item(&mut self, arg: &PortDeclarationItem) -> Result<(), ParolError> {
+        if let HandlerPoint::Before = self.point {
+            if let PortDeclarationItemGroup::DirectionArrayType(x) =
+                arg.port_declaration_item_group.as_ref()
+            {
+                if let Direction::Inout(_) = x.direction.as_ref() {
+                    let r#type = &x.array_type;
+                    let is_tri = r#type
+                        .scalar_type
+                        .scalar_type_list
+                        .iter()
+                        .any(|x| matches!(x.type_modifier.as_ref(), TypeModifier::Tri(_)));
+
+                    if !is_tri {
+                        self.errors.push(AnalyzerError::missing_tri(
+                            self.text,
+                            &r#type.as_ref().into(),
+                        ));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
     fn direction(&mut self, arg: &Direction) -> Result<(), ParolError> {
         if let HandlerPoint::Before = self.point {
             match arg {

--- a/crates/analyzer/src/symbol.rs
+++ b/crates/analyzer/src/symbol.rs
@@ -324,7 +324,7 @@ pub enum TypeKind {
     UserDefined(Vec<StrId>),
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum TypeModifier {
     Tri,
     Signed,

--- a/crates/analyzer/src/tests.rs
+++ b/crates/analyzer/src/tests.rs
@@ -351,6 +351,19 @@ fn missing_reset_statement() {
 }
 
 #[test]
+fn missing_tri() {
+    let code = r#"
+    module ModuleA (
+        x: inout logic,
+    ) {
+    }
+    "#;
+
+    let errors = analyze(code);
+    assert!(matches!(errors[0], AnalyzerError::MissingTri { .. }));
+}
+
+#[test]
 fn too_large_enum_variant() {
     let code = r#"
     module ModuleA {

--- a/crates/emitter/src/emitter.rs
+++ b/crates/emitter/src/emitter.rs
@@ -3,6 +3,7 @@ use veryl_analyzer::attribute::Attribute as Attr;
 use veryl_analyzer::attribute_table;
 use veryl_analyzer::namespace::Namespace;
 use veryl_analyzer::symbol::SymbolKind;
+use veryl_analyzer::symbol::TypeModifier as SymTypeModifier;
 use veryl_analyzer::symbol_table::{self, SymbolPath, SymbolPathNamespace};
 use veryl_analyzer::{msb_table, namespace_table};
 use veryl_metadata::{Build, BuiltinType, ClockType, Format, Metadata, ResetType};
@@ -523,11 +524,6 @@ impl VerylWalker for Emitter {
         } else {
             self.veryl_token(&arg.comma_token);
         }
-    }
-
-    /// Semantic action for non-terminal 'Assign'
-    fn assign(&mut self, arg: &Assign) {
-        self.veryl_token(&arg.assign_token.replace("always_comb"));
     }
 
     /// Semantic action for non-terminal 'F32'
@@ -1498,6 +1494,13 @@ impl VerylWalker for Emitter {
 
     /// Semantic action for non-terminal 'LetDeclaration'
     fn let_declaration(&mut self, arg: &LetDeclaration) {
+        let is_tri = arg
+            .array_type
+            .scalar_type
+            .scalar_type_list
+            .iter()
+            .any(|x| matches!(x.type_modifier.as_ref(), TypeModifier::Tri(_)));
+
         self.scalar_type(&arg.array_type.scalar_type);
         self.space(1);
         self.identifier(&arg.identifier);
@@ -1507,7 +1510,11 @@ impl VerylWalker for Emitter {
         }
         self.str(";");
         self.newline();
-        self.str("always_comb");
+        if is_tri {
+            self.str("assign");
+        } else {
+            self.str("always_comb");
+        }
         self.space(1);
         self.str(&arg.identifier.identifier_token.to_string());
         self.space(1);
@@ -1704,7 +1711,28 @@ impl VerylWalker for Emitter {
 
     /// Semantic action for non-terminal 'AssignDeclaration'
     fn assign_declaration(&mut self, arg: &AssignDeclaration) {
-        self.assign(&arg.assign);
+        let emit_assign =
+            if let Ok(symbol) = symbol_table::resolve(arg.hierarchical_identifier.as_ref()) {
+                match &symbol.found.kind {
+                    SymbolKind::Variable(x) => x.r#type.modifier.contains(&SymTypeModifier::Tri),
+                    SymbolKind::Port(x) => {
+                        if let Some(ref x) = x.r#type {
+                            x.modifier.contains(&SymTypeModifier::Tri)
+                        } else {
+                            false
+                        }
+                    }
+                    _ => false,
+                }
+            } else {
+                // External symbols may be tri-state
+                true
+            };
+        if emit_assign {
+            self.assign(&arg.assign);
+        } else {
+            self.token(&arg.assign.assign_token.replace("always_comb"));
+        }
         self.space(1);
         self.hierarchical_identifier(&arg.hierarchical_identifier);
         self.space(1);

--- a/crates/parser/src/veryl_token.rs
+++ b/crates/parser/src/veryl_token.rs
@@ -245,6 +245,76 @@ impl From<&AlwaysFfDeclaration> for TokenRange {
     }
 }
 
+impl From<&FixedType> for TokenRange {
+    fn from(value: &FixedType) -> Self {
+        let beg = match value {
+            FixedType::U32(x) => x.u32.u32_token.token,
+            FixedType::U64(x) => x.u64.u64_token.token,
+            FixedType::I32(x) => x.i32.i32_token.token,
+            FixedType::I64(x) => x.i64.i64_token.token,
+            FixedType::F32(x) => x.f32.f32_token.token,
+            FixedType::F64(x) => x.f64.f64_token.token,
+            FixedType::Strin(x) => x.strin.string_token.token,
+        };
+        let end = beg;
+        TokenRange { beg, end }
+    }
+}
+
+impl From<&VariableType> for TokenRange {
+    fn from(value: &VariableType) -> Self {
+        let mut range = match &*value.variable_type_group {
+            VariableTypeGroup::Logic(x) => {
+                let beg = x.logic.logic_token.token;
+                let end = beg;
+                TokenRange { beg, end }
+            }
+            VariableTypeGroup::Bit(x) => {
+                let beg = x.bit.bit_token.token;
+                let end = beg;
+                TokenRange { beg, end }
+            }
+            VariableTypeGroup::ScopedIdentifier(x) => x.scoped_identifier.as_ref().into(),
+        };
+
+        if let Some(ref x) = value.variable_type_opt {
+            range.end = x.width.r_angle.r_angle_token.token;
+        }
+
+        range
+    }
+}
+
+impl From<&ScalarType> for TokenRange {
+    fn from(value: &ScalarType) -> Self {
+        let mut range: TokenRange = match &*value.scalar_type_group {
+            ScalarTypeGroup::VariableType(x) => x.variable_type.as_ref().into(),
+            ScalarTypeGroup::FixedType(x) => x.fixed_type.as_ref().into(),
+        };
+
+        if let Some(x) = value.scalar_type_list.first() {
+            range.beg = match &*x.type_modifier {
+                TypeModifier::Tri(x) => x.tri.tri_token.token,
+                TypeModifier::Signed(x) => x.r#signed.signed_token.token,
+            };
+        }
+
+        range
+    }
+}
+
+impl From<&ArrayType> for TokenRange {
+    fn from(value: &ArrayType) -> Self {
+        let mut range: TokenRange = value.scalar_type.as_ref().into();
+
+        if let Some(ref x) = value.array_type_opt {
+            range.end = x.array.r_bracket.r_bracket_token.token;
+        }
+
+        range
+    }
+}
+
 #[derive(Debug, Clone)]
 pub struct VerylToken {
     pub token: Token,

--- a/testcases/sv/04_module.sv
+++ b/testcases/sv/04_module.sv
@@ -25,7 +25,7 @@ module veryl_testcase_Module04 #(
     always_comb _ddd = 1;
 
     always_comb bb  = 0;
-    always_comb bbb = 0;
+    assign bbb = 0;
 endmodule
 
 interface veryl_testcase_Interface04;

--- a/testcases/sv/22_type_modifier.sv
+++ b/testcases/sv/22_type_modifier.sv
@@ -2,9 +2,9 @@ module veryl_testcase_Module22;
     logic     signed [10-1:0] _a;
     always_comb _a = 1;
     tri logic signed [10-1:0] _b;
-    always_comb _b = 1;
+    assign _b = 1;
     tri logic signed [10-1:0] _c;
-    always_comb _c = 1;
+    assign _c = 1;
     logic            [10-1:0] _d;
     always_comb _d = 1;
 endmodule

